mobile/lib/shared/services/hash.service.dart

```diff
@@ -0,0 +1,183 @@
+    AssetPathEntity album, {
+    int start = 0,
+    int end = 0x7fffffffffffffff,
+    Set<String>? excludedAssets,
+  }) async {
+    final entities = await album.getAssetListRange(start: start, end: end);
+    final filtered = excludedAssets == null
+        ? entities
+        : entities.where((e) => !excludedAssets.contains(e.id)).toList();
+    return hashAssets(filtered);
+  }
+
+  Future<List<int>?> hashAsset(Asset a) async {
+    assert(a.isLocal, "can only get hash of local assets");
+    final deviceAsset = await (Platform.isAndroid
+        ? _db.androidDeviceAssets.get(a.localId!.toInt())
+        : _db.iOSDeviceAssets.getById(a.localId!));
+    if (deviceAsset != null) return deviceAsset.hash;
+    final file = await a.local!.originFile;
+    if (file == null) {
+      _log.warning("Failed to get file for asset ${a.id}, skipping");
+      return null;
+    }
+    final hash =
+        await (Platform.isAndroid ? _hashFile(file) : _hashAssetCrypto(file));
+    if (hash == null) {
+      _log.warning("Failed to hash file ${file.path} for asset ${a.id}");
+      return null;
+    }
+    if (Platform.isAndroid) {
+      final da = AndroidDeviceAsset(id: a.localId!.toInt(), hash: hash);
+      await _db.writeTxn(() => _db.androidDeviceAssets.put(da));
+    } else {
+      final da = IOSDeviceAsset(id: a.localId!, hash: hash);
+      await _db.writeTxn(() => _db.iOSDeviceAssets.put(da));
+    }
+    return hash;
+  }
+
+  Future<List<Asset>> hashAssets(List<AssetEntity> assets) async {
+    final Stopwatch totalTime = Stopwatch()..start();
+    const int batchFileCount = 128;
+    const int batchDataSize = 1024 * 1024 * 1024; // 1GB
+    int totalBytes = 0;
+    int totalFiles = 0;
+    final List<DeviceAsset?> hashes;
+    final Stopwatch sw = Stopwatch()..start();
+    final List<DeviceAsset> toAdd = [];
+    final ids = assets
+        .map(Platform.isAndroid ? (a) => a.id.toInt() : (a) => a.id)
+        .toList();
+    hashes = await (Platform.isAndroid
+        ? _db.androidDeviceAssets.getAll(ids.cast())
+        : _db.iOSDeviceAssets.getAllById(ids.cast()));
+    debugPrint("Checking hash DB took ${sw.elapsedMilliseconds}ms");
+    sw.reset();
+    int bytes = 0;
+    final List<String> toHash = [];
+    for (int i = 0; i < assets.length; i++) {
+      if (hashes[i] != null) continue;
+      final file = await assets[i].originFile;
+      if (file == null) {
+        _log.warning("Failed to get file for asset ${assets[i].id}, skipping");
+        continue;
+      }
+      bytes += file.lengthSync();
+      toHash.add(file.path);
+      final da = Platform.isAndroid
+          ? AndroidDeviceAsset(id: ids[i] as int, hash: const [])
+          : IOSDeviceAsset(id: ids[i] as String, hash: const []);
+      toAdd.add(da);
+      totalFiles++;
+      hashes[i] = da;
+      if (toHash.length == batchFileCount ||
+          bytes >= batchDataSize ||
+          (i + 1 == assets.length && toHash.isNotEmpty)) {
```

Comment:
This line seems buggy. There are several situations where the for loop would `continue` above on potentially the last item, and this code would never execute. It needs to be run here and after the for loop exits.

Reply:
good catch!